internal/integration/rhp/v4/rhp4_test.go

```diff
@@ -204,7 +204,284 @@ func TestFormContract(t *testing.T) {
+		} else if !known {
+			t.Fatal("expected transaction set to be known")
+		}
+
+		sigHash := cm.TipState().ContractSigHash(refreshResult.Contract.Revision)
+		if !renterKey.PublicKey().VerifyHash(sigHash, refreshResult.Contract.Revision.RenterSignature) {
+			t.Fatal("renter signature verification failed")
+		} else if !hostKey.PublicKey().VerifyHash(sigHash, refreshResult.Contract.Revision.HostSignature) {
+			t.Fatal("host signature verification failed")
+		}
+
+		rs, err := rhp4.RPCLatestRevision(context.Background(), transport, revision.ID)
+		if err != nil {
+			t.Fatal(err)
+		} else if !rs.Renewed {
+			t.Fatal("expected contract to be renewed")
+		} else if rs.Revisable {
+			t.Fatal("expected contract to not be revisable")
+		}
+	})
+
+	t.Run("valid refresh - partial funding", func(t *testing.T) {
+		revision := formContractUploadSector(t, types.Siacoins(10), types.Siacoins(20), types.Siacoins(1))
+
+		newAllowance := types.Siacoins(20)
+		newCollateral := types.Siacoins(40)
+		// refresh the contract
+		refreshResult, err := rhp4.RPCRefreshContractPartialRollover(context.Background(), transport, cm, fundAndSign, cm.TipState(), settings.Prices, revision.Revision, proto4.RPCRefreshContractParams{
+			ContractID: revision.ID,
+			Allowance:  newAllowance,
+			Collateral: newCollateral,
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		assertRefreshValues(t, settings.Prices, revision.Revision, refreshResult.Contract.Revision, newAllowance, newCollateral)
+		assertRollover(t, settings.Prices, revision.Revision, refreshResult.RenewalSet.Transactions, newAllowance, newCollateral)
+
+		// verify the transaction set is valid
+		if known, err := cm.AddV2PoolTransactions(refreshResult.RenewalSet.Basis, refreshResult.RenewalSet.Transactions); err != nil {
+			t.Fatal(err)
+		} else if !known {
+			t.Fatal("expected transaction set to be known")
+		}
+
+		sigHash := cm.TipState().ContractSigHash(refreshResult.Contract.Revision)
+		if !renterKey.PublicKey().VerifyHash(sigHash, refreshResult.Contract.Revision.RenterSignature) {
+			t.Fatal("renter signature verification failed")
+		} else if !hostKey.PublicKey().VerifyHash(sigHash, refreshResult.Contract.Revision.HostSignature) {
+			t.Fatal("host signature verification failed")
+		}
+
+		rs, err := rhp4.RPCLatestRevision(context.Background(), transport, revision.ID)
+		if err != nil {
+			t.Fatal(err)
+		} else if !rs.Renewed {
+			t.Fatal("expected contract to be renewed")
+		} else if rs.Revisable {
+			t.Fatal("expected contract to not be revisable")
+		}
+	})
+
+	t.Run("valid refresh - no renter funding", func(t *testing.T) {
+		expirationHeight := cm.Tip().Height + 50 + proto4.ProofWindow
+		sectorCost := settings.Prices.RPCAppendSectorsCost(1, expirationHeight-settings.Prices.TipHeight)
+		accountFunding := types.Siacoins(10).Sub(sectorCost.RenterCost())
+		revision := formContractUploadSector(t, types.Siacoins(10), types.Siacoins(20), accountFunding)
+
+		newAllowance := types.Siacoins(10)
```

Comment:
(nit) I'm not sure I understand "no renter funding" correctly but assuming you mean that the allowance of the renter stays the same, this should probably be `newAllowance := revision.Revision.RemainingAllowance()`?

Reply:
It means the renter does not need to fund the transaction. It's copied from coreutils, so I'm not terribly concerned.